File: app-android/src/main/kotlin/org/kopiaKt/app/ui/screens/repositoryconnect/RepositoryConnectScreen.kt
package org.kopiaKt.app.ui.screens.repositoryconnect

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import org.kopiaKt.app.domain.model.StorageType

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RepositoryConnectScreen(
    onConnected: () -> Unit,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
    viewModel: RepositoryConnectViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsState()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Connect Repository") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, "Back")
                    }
                }
            )
        }
    ) { padding ->
        Column(
            modifier = modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(16.dp)
        ) {
            // Storage type tabs
            val storageTypes = listOf(StorageType.LOCAL_FILESYSTEM, StorageType.S3, StorageType.WEBDAV, StorageType.SFTP)
            val selectedIndex = storageTypes.indexOf(uiState.selectedStorageType).coerceAtLeast(0)

            ScrollableTabRow(selectedTabIndex = selectedIndex) {
                storageTypes.forEachIndexed { index, type ->
                    Tab(
                        selected = selectedIndex == index,
                        onClick = { viewModel.updateStorageType(type) },
                        text = { Text(if (type == StorageType.LOCAL_FILESYSTEM) "Local" else type.name) }
                    )
                }
            }

            Spacer(modifier = Modifier.height(24.dp))

            // Storage-specific form fields
            when (uiState.selectedStorageType) {
                StorageType.LOCAL_FILESYSTEM -> LocalForm(
                    state = uiState.localConfig,
                    onUpdate = viewModel::updateLocalConfig
                )
                StorageType.S3 -> S3Form(
                    state = uiState.s3Config,
                    onUpdate = viewModel::updateS3Config
                )
                StorageType.WEBDAV -> WebDavForm(
                    state = uiState.webDavConfig,
                    onUpdate = viewModel::updateWebDavConfig
                )
                StorageType.SFTP -> SftpForm(
                    state = uiState.sftpConfig,
                    onUpdate = viewModel::updateSftpConfig
                )
                else -> {
                    Text("Storage type not supported yet")
                }
            }

            Spacer(modifier = Modifier.height(24.dp))

            // Password field
            OutlinedTextField(
                value = uiState.password,
                onValueChange = viewModel::updatePassword,
                label = { Text("Repository Password") },
                visualTransformation = PasswordVisualTransformation(),
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag("password_field"),
                singleLine = true
            )

            Spacer(modifier = Modifier.height(8.dp))

            // Remember password checkbox
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth()
            ) {
                Checkbox(
                    checked = uiState.savePassword,
                    onCheckedChange = viewModel::updateSavePassword
                )
                Text("Remember password")
            }

            // Error message
            uiState.error?.let { error ->
                Spacer(modifier = Modifier.height(16.dp))
                Text(
                    text = error,
                    color = MaterialTheme.colorScheme.error,
                    style = MaterialTheme.typography.bodyMedium
                )
            }

            Spacer(modifier = Modifier.height(24.dp))

            // Connect button
            Button(
                onClick = { viewModel.connect(onConnected) },
                enabled = !uiState.isConnecting,
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag("connect_button")
            ) {
                if (uiState.isConnecting) {
                    CircularProgressIndicator(
                        modifier = Modifier.size(24.dp),
                        color = MaterialTheme.colorScheme.onPrimary
                    )
                } else {
                    Text("Connect")
                }
            }
        }
    }
}

@Composable
private fun LocalForm(
    state: LocalFormState,
    onUpdate: (path: String) -> Unit
) {
    OutlinedTextField(
        value = state.path,
        onValueChange = onUpdate,
        label = { Text("Repository Path") },
        placeholder = { Text("/sdcard/kopia_repo") },
        modifier = Modifier
            .fillMaxWidth()
            .testTag("repo_path_field"),
        singleLine = true,
        keyboardOptions = KeyboardOptions(
            autoCorrect = false,
            keyboardType = KeyboardType.Uri
        )
    )
    Spacer(modifier = Modifier.height(8.dp))
    Text(
        text = "Enter the full path to the Kopia repository directory on the device.",
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant
    )
}

@Composable
private fun S3Form(
    state: S3FormState,
    onUpdate: (bucket: String, endpoint: String, region: String, accessKeyId: String) -> Unit
) {
    OutlinedTextField(
        value = state.bucket,
        onValueChange = { onUpdate(it, state.endpoint, state.region, state.accessKeyId) },
        label = { Text("Bucket") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
    Spacer(modifier = Modifier.height(8.dp))
    OutlinedTextField(
        value = state.endpoint,
        onValueChange = { onUpdate(state.bucket, it, state.region, state.accessKeyId) },
        label = { Text("Endpoint") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
    Spacer(modifier = Modifier.height(8.dp))
    OutlinedTextField(
        value = state.region,
        onValueChange = { onUpdate(state.bucket, state.endpoint, it, state.accessKeyId) },
        label = { Text("Region") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
    Spacer(modifier = Modifier.height(8.dp))
    OutlinedTextField(
        value = state.accessKeyId,
        onValueChange = { onUpdate(state.bucket, state.endpoint, state.region, it) },
        label = { Text("Access Key ID") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
}

@Composable
private fun WebDavForm(
    state: WebDavFormState,
    onUpdate: (url: String, username: String) -> Unit
) {
    OutlinedTextField(
        value = state.url,
        onValueChange = { onUpdate(it, state.username) },
        label = { Text("WebDAV URL") },
        placeholder = { Text("https://example.com/dav/") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
    Spacer(modifier = Modifier.height(8.dp))
    OutlinedTextField(
        value = state.username,
        onValueChange = { onUpdate(state.url, it) },
        label = { Text("Username") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
}

@Composable
private fun SftpForm(
    state: SftpFormState,
    onUpdate: (host: String, port: Int, username: String, path: String) -> Unit
) {
    OutlinedTextField(
        value = state.host,
        onValueChange = { onUpdate(it, state.port, state.username, state.path) },
        label = { Text("Host") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
    Spacer(modifier = Modifier.height(8.dp))
    OutlinedTextField(
        value = state.port.toString(),
        onValueChange = { onUpdate(state.host, it.toIntOrNull() ?: 22, state.username, state.path) },
        label = { Text("Port") },
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
    Spacer(modifier = Modifier.height(8.dp))
    OutlinedTextField(
        value = state.username,
        onValueChange = { onUpdate(state.host, state.port, it, state.path) },
        label = { Text("Username") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
    Spacer(modifier = Modifier.height(8.dp))
    OutlinedTextField(
        value = state.path,
        onValueChange = { onUpdate(state.host, state.port, state.username, it) },
        label = { Text("Path") },
        placeholder = { Text("/path/to/repository") },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true
    )
}

File: flutter_ui/lib/screens/repository_connect_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../bridge/kopia_bridge.g.dart';
import '../providers/repository_connect_provider.dart';

/// Screen for connecting to a Kopia repository.
/// Supports Local filesystem, S3, WebDAV, and SFTP storage types.
class RepositoryConnectScreen extends ConsumerWidget {
  const RepositoryConnectScreen({
    super.key,
    required this.onConnected,
    required this.onBack,
  });

  final VoidCallback onConnected;
  final VoidCallback onBack;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final state = ref.watch(repositoryConnectProvider);
    final notifier = ref.read(repositoryConnectProvider.notifier);

    return Scaffold(
      appBar: AppBar(
        title: const Text('Connect Repository'),
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: onBack,
        ),
      ),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            // Storage type selector
            _StorageTypeSelector(
              selectedType: state.selectedStorageType,
              onTypeSelected: notifier.setStorageType,
            ),
            const SizedBox(height: 24),

            // Storage-specific form
            _buildStorageForm(context, state, notifier),
            const SizedBox(height: 24),

            // Password field
            TextField(
              key: const Key('password_field'),
              onChanged: notifier.setPassword,
              obscureText: true,
              autocorrect: false,
              enableSuggestions: false,
              decoration: const InputDecoration(
                labelText: 'Repository Password',
                border: OutlineInputBorder(),
              ),
            ),
            const SizedBox(height: 8),

            // Remember password checkbox
            Row(
              children: [
                Checkbox(
                  value: state.savePassword,
                  onChanged: (value) => notifier.setSavePassword(value ?? true),
                ),
                const Text('Remember password'),
              ],
            ),

            // Error message
            if (state.error != null) ...[
              const SizedBox(height: 16),
              Text(
                state.error!,
                style: TextStyle(
                  color: Theme.of(context).colorScheme.error,
                ),
              ),
            ],
            const SizedBox(height: 24),

            // Connect button
            FilledButton(
              key: const Key('connect_button'),
              onPressed: state.isConnecting
                  ? null
                  : () async {
                      final success = await notifier.connect();
                      if (success) {
                        onConnected();
                      }
                    },
              child: state.isConnecting
                  ? const SizedBox(
                      height: 24,
                      width: 24,
                      child: CircularProgressIndicator(strokeWidth: 2),
                    )
                  : const Text('Connect'),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildStorageForm(
    BuildContext context,
    RepositoryConnectState state,
    RepositoryConnectNotifier notifier,
  ) {
    switch (state.selectedStorageType) {
      case StorageType.localFilesystem:
        return _LocalFilesystemForm(
          path: state.localPath,
          onPathChanged: notifier.setLocalPath,
        );
      case StorageType.s3:
        return _S3Form(
          bucket: state.s3Bucket,
          endpoint: state.s3Endpoint,
          region: state.s3Region,
          accessKeyId: state.s3AccessKeyId,
          onBucketChanged: notifier.setS3Bucket,
          onEndpointChanged: notifier.setS3Endpoint,
          onRegionChanged: notifier.setS3Region,
          onAccessKeyIdChanged: notifier.setS3AccessKeyId,
        );
      case StorageType.webdav:
        return _WebDavForm(
          url: state.webDavUrl,
          username: state.webDavUsername,
          onUrlChanged: notifier.setWebDavUrl,
          onUsernameChanged: notifier.setWebDavUsername,
        );
      case StorageType.sftp:
        return _SftpForm(
          host: state.sftpHost,
          port: state.sftpPort,
          username: state.sftpUsername,
          path: state.sftpPath,
          onHostChanged: notifier.setSftpHost,
          onPortChanged: notifier.setSftpPort,
          onUsernameChanged: notifier.setSftpUsername,
          onPathChanged: notifier.setSftpPath,
        );
      default:
        return Text(
          'Storage type not supported yet',
          style: Theme.of(context).textTheme.bodyMedium,
        );
    }
  }
}

/// Tab selector for storage type
class _StorageTypeSelector extends StatelessWidget {
  const _StorageTypeSelector({
    required this.selectedType,
    required this.onTypeSelected,
  });

  final StorageType selectedType;
  final ValueChanged<StorageType> onTypeSelected;

  static const _supportedTypes = [
    StorageType.localFilesystem,
    StorageType.s3,
    StorageType.webdav,
    StorageType.sftp,
  ];

  String _getLabel(StorageType type) {
    switch (type) {
      case StorageType.localFilesystem:
        return 'Local';
      case StorageType.s3:
        return 'S3';
      case StorageType.webdav:
        return 'WebDAV';
      case StorageType.sftp:
        return 'SFTP';
      default:
        return type.name;
    }
  }

  @override
  Widget build(BuildContext context) {
    final selectedIndex = _supportedTypes.indexOf(selectedType);

    return DefaultTabController(
      length: _supportedTypes.length,
      initialIndex: selectedIndex >= 0 ? selectedIndex : 0,
      child: TabBar(
        isScrollable: true,
        onTap: (index) => onTypeSelected(_supportedTypes[index]),
        tabs: _supportedTypes.map((type) => Tab(text: _getLabel(type))).toList(),
      ),
    );
  }
}

/// Form for local filesystem connection
class _LocalFilesystemForm extends StatelessWidget {
  const _LocalFilesystemForm({
    required this.path,
    required this.onPathChanged,
  });

  final String path;
  final ValueChanged<String> onPathChanged;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        TextField(
          key: const Key('repo_path_field'),
          controller: TextEditingController(text: path)
            ..selection = TextSelection.collapsed(offset: path.length),
          onChanged: onPathChanged,
          autocorrect: false,
          enableSuggestions: false,
          keyboardType: TextInputType.url,
          decoration: const InputDecoration(
            labelText: 'Repository Path',
            hintText: '/sdcard/kopia_repo',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        Text(
          'Enter the full path to the Kopia repository directory on the device.',
          style: Theme.of(context).textTheme.bodySmall?.copyWith(
                color: Theme.of(context).colorScheme.onSurfaceVariant,
              ),
        ),
      ],
    );
  }
}

/// Form for S3 connection
class _S3Form extends StatelessWidget {
  const _S3Form({
    required this.bucket,
    required this.endpoint,
    required this.region,
    required this.accessKeyId,
    required this.onBucketChanged,
    required this.onEndpointChanged,
    required this.onRegionChanged,
    required this.onAccessKeyIdChanged,
  });

  final String bucket;
  final String endpoint;
  final String region;
  final String accessKeyId;
  final ValueChanged<String> onBucketChanged;
  final ValueChanged<String> onEndpointChanged;
  final ValueChanged<String> onRegionChanged;
  final ValueChanged<String> onAccessKeyIdChanged;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        TextField(
          controller: TextEditingController(text: bucket)
            ..selection = TextSelection.collapsed(offset: bucket.length),
          onChanged: onBucketChanged,
          decoration: const InputDecoration(
            labelText: 'Bucket',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        TextField(
          controller: TextEditingController(text: endpoint)
            ..selection = TextSelection.collapsed(offset: endpoint.length),
          onChanged: onEndpointChanged,
          decoration: const InputDecoration(
            labelText: 'Endpoint',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        TextField(
          controller: TextEditingController(text: region)
            ..selection = TextSelection.collapsed(offset: region.length),
          onChanged: onRegionChanged,
          decoration: const InputDecoration(
            labelText: 'Region',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        TextField(
          controller: TextEditingController(text: accessKeyId)
            ..selection = TextSelection.collapsed(offset: accessKeyId.length),
          onChanged: onAccessKeyIdChanged,
          decoration: const InputDecoration(
            labelText: 'Access Key ID',
            border: OutlineInputBorder(),
          ),
        ),
      ],
    );
  }
}

/// Form for WebDAV connection
class _WebDavForm extends StatelessWidget {
  const _WebDavForm({
    required this.url,
    required this.username,
    required this.onUrlChanged,
    required this.onUsernameChanged,
  });

  final String url;
  final String username;
  final ValueChanged<String> onUrlChanged;
  final ValueChanged<String> onUsernameChanged;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        TextField(
          controller: TextEditingController(text: url)
            ..selection = TextSelection.collapsed(offset: url.length),
          onChanged: onUrlChanged,
          decoration: const InputDecoration(
            labelText: 'WebDAV URL',
            hintText: 'https://example.com/dav/',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        TextField(
          controller: TextEditingController(text: username)
            ..selection = TextSelection.collapsed(offset: username.length),
          onChanged: onUsernameChanged,
          decoration: const InputDecoration(
            labelText: 'Username',
            border: OutlineInputBorder(),
          ),
        ),
      ],
    );
  }
}

/// Form for SFTP connection
class _SftpForm extends StatelessWidget {
  const _SftpForm({
    required this.host,
    required this.port,
    required this.username,
    required this.path,
    required this.onHostChanged,
    required this.onPortChanged,
    required this.onUsernameChanged,
    required this.onPathChanged,
  });

  final String host;
  final int port;
  final String username;
  final String path;
  final ValueChanged<String> onHostChanged;
  final ValueChanged<int> onPortChanged;
  final ValueChanged<String> onUsernameChanged;
  final ValueChanged<String> onPathChanged;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        TextField(
          controller: TextEditingController(text: host)
            ..selection = TextSelection.collapsed(offset: host.length),
          onChanged: onHostChanged,
          decoration: const InputDecoration(
            labelText: 'Host',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        TextField(
          controller: TextEditingController(text: port.toString())
            ..selection = TextSelection.collapsed(offset: port.toString().length),
          onChanged: (value) {
            final parsed = int.tryParse(value);
            if (parsed != null) {
              onPortChanged(parsed);
            }
          },
          keyboardType: TextInputType.number,
          decoration: const InputDecoration(
            labelText: 'Port',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        TextField(
          controller: TextEditingController(text: username)
            ..selection = TextSelection.collapsed(offset: username.length),
          onChanged: onUsernameChanged,
          decoration: const InputDecoration(
            labelText: 'Username',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 8),
        TextField(
          controller: TextEditingController(text: path)
            ..selection = TextSelection.collapsed(offset: path.length),
          onChanged: onPathChanged,
          decoration: const InputDecoration(
            labelText: 'Path',
            hintText: '/path/to/repository',
            border: OutlineInputBorder(),
          ),
        ),
      ],
    );
  }
}
